File: c++/year_2023/day_1/part_1/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <sstream>

using namespace std;

vector<string>* read_from_file(string fileName) {
    vector<string>* v = new vector<string>;
    string str;

    ifstream fin(fileName);
    while (fin >> str) {
        v -> push_back(str);
    }
    fin.close();

    return v;
}

int get_calibration_value(string *str) {
    char last, first = '\0';

    for (int i = 0; i < str -> length(); i++) {
        if (isdigit(str -> at(i))) {
            if (first == '\0') {
                first = str -> at(i);
            }
            last = str -> at(i);
        }
    }

    stringstream ss;
    ss << first << last;
    return stoi(ss.str());
}

int sum_calibration_values(string fileName) {
    vector<string>* v = read_from_file(fileName);

    int sum = 0;
    for (int i = 0; i < v -> size(); i++) {
        sum += get_calibration_value(&v -> at(i));
    }

    return sum;
}

int main() {
    cout << "---Example---" << endl << sum_calibration_values("example.txt") << endl;
    cout << "---Input---" << endl << sum_calibration_values("input.txt") << endl;

    return 0;
}

File: python/year_2023/day_1/part_1/main.py
import os


def process_calibration_words(filename: str) -> list[str]:
    with open(os.path.join(os.path.dirname(__file__), filename), encoding="utf-8") as f:
        return list(filter(None, f.read().split("\n")))


def get_calibration_value(calibrationWord: str) -> int:
    first, last = None, None

    for letter in calibrationWord:
        if letter.isdigit():
            if first is None:
                first = letter
            last = letter

    return int("%s%s" % (first, last))

def main(filename: str) -> int:
    calibrationWords = process_calibration_words(filename)
    
    return sum(map(get_calibration_value, calibrationWords))



if __name__ == "__main__":
    print("---Example---")
    print(main("example.txt"))
    
    print("---Input---")
    print(main("input.txt"))
